### ISSUE TITLE
Login JavaScript Error

### ISSUE THREAD

**Issue by jajmo** [sharplab/spring-security-webauthn#2]
Upon trying to login to the site with a registered token, I get the following JavaScript console error:

```DOMException: "An attempt was made to use an object that is not, or is no longer, usable"```

This is thrown from `loginViewModel.js:48`. This occurred when attempting to login with a registered token in Firefox. I spent some time trying to debug this to potentially find a solution, to no avail.

On top of that, when trying to login with Chrome, the token doesn't even blink.

**Comment by ynojima** [comment]
Could you restart your browser and try again? It seems browser throws InvalidStateError.
Current sample application doesn't have enough debug outputs, I cannot destruct the issue.
I'm re-implementing sample application with Angular now, and I'll solve the issue there.

**Comment by jajmo** [comment]
I've restarted the computer several times since the issue started happening.

I guess I'll wait for the Angular version to be ready. Do you have an ETA on that?

**Comment by ynojima** [comment]
Thank you for your patience. It will be come within one month or two.

**Comment by jajmo** [comment]
Awesome. Do you have a ETA for this getting merged into Spring Security? Or is that waiting on Angular too.

**Comment by ynojima** [comment]
It's basically up on spring-security side's decision, but I also want to confirm spring-security-webauthn works with Safari, which is the last major browser not supporting WebAuthn for now.
They are developing the WebAuthn feature now (ref. https://bugs.webkit.org/show_bug.cgi?id=181943).
So, there is no ETA.

**Comment by jajmo** [comment]
I'm still having the same issue. However, the frontend looks exactly the same as it did before.

I'm executing it behind an nginx reverse proxy. The jar being executed after building is `spring-security-webauthn/sample/build/libs/spring-security-webauthn-sample.jar`

Please let me know if this is incorrect.

**Comment by ynojima** [comment]
It seems you are running older version.
Hove you pulled and rebuilt latest master? 

The jar you should use will be generated at 
`samples/javaconfig/webauthn/spa/build/spring-security-webauthn-sample-spa.jar`.

It has a login view looks like this:
https://github.com/sharplab/spring-security-webauthn/blob/b07422d58cb1216746a969c31e084dde7ab8b305/docs/src/reference/asciidoc/en/images/login.png

**Comment by jajmo** [comment]
Yes, I have it running like that now, but it's only accessible through localhost. When I try to access it through the external domain, it redirects me to localhost, which is incorrect.

**Comment by ynojima** [comment]
Have you resolved the jar issue? Good.

When are you redirected? Immediately after requesting login page? or after pressing the login button?

**Comment by jajmo** [comment]
Immediately after requesting the login page it redirects to localhost/angular/login. I can't find anywhere to configure the proper domain name.

**Comment by ynojima** [comment]
Hmm. It's by this code(https://github.com/sharplab/spring-security-webauthn/blob/57231892f1527903ae6d71f588bc9a013413b6c5/samples/javaconfig/webauthn/spa/src/main/java/net/sharplab/springframework/security/webauthn/sample/app/web/RootController.java#L17).
But in my understanding, redirect is done based on the domain that spring aware. Does your reverse proxy provide "X-Forwarded-For" HTTP header?

**Comment by ynojima** [comment]
spring-security-webauthn-sample-spa awares ""X-Forwarded-For" and ""X-Forwarded-Proto" by this config
https://github.com/sharplab/spring-security-webauthn/blob/b07422d58cb1216746a969c31e084dde7ab8b305/samples/javaconfig/webauthn/spa/src/main/resources/application.yml#L27

**Comment by ynojima** [comment]
Here is the sample application hosted online behind proxy:
https://login.without.pw/angular/signup

**Comment by jajmo** [comment]
Hm, maybe it's an issue with my nginx configuration then. I have it set to add X-Forwarded-For, X-Forwarded-Proto, and X-Forwarded-Port.

**Comment by jajmo** [comment]
Figured it out. It redirects properly when the `Host` header is set.

**Comment by ynojima** [comment]
Oh, yes, `Host` header is the one to be set.

So, is your original issue `DOMException: "An attempt was made to use an object that is not, or is no longer, usable"` reproduced with new Angular app?

**Comment by jajmo** [comment]
No, but now in Chrome when I click "Device Login" none of my authenticators blink. The popup in Chrome comes up asking for me to activate one of them, but no matter which I try it doesn't work.

**Comment by ynojima** [comment]
Sorry for misleading. As I've written in another issue (https://github.com/sharplab/spring-security-webauthn/issues/3), for WebAuthn for second-factor scenario, you should use "Login" button instead of "Device Login" button. "Device Login" button is for password-less login, which is not supported by Chrome and Firefox. Only Microsoft Edge supports it.

**Comment by jajmo** [comment]
Again, both Firefox and Chrome have supported passwordless authentication for quite some time. Microsoft Edge was the last browser to support it, adding support in the latest Windows Insider Preview build in October I believe.

Chrome: https://www.tomshardware.com/news/chrome-67-pwa-passwordless-login,37144.html
`Chrome 67 brings support for the Web Authentication API, which will allow users to log in to websites without needing a password through a private-public key cryptography system.`

Firefox: https://www.digitaltrends.com/computing/firefox-60-supports-webauthn-password-free-logins/
`Firefox 60 is the first browser to support password-free internet logins`

Please correct me if I'm misunderstanding what you're saying.

**Comment by ynojima** [comment]
I've answered in the https://github.com/sharplab/spring-security-webauthn/issues/3#issuecomment-443781206 about the Firefox, let's continue the discussion in this issue #2.

**Comment by jajmo** [comment]
That post is from January. It is now December, and Firefox is on version 63. J.C. clearly states WebAuthN (passwordless login) was due to ship in Firefox 60. Which it did. If you go to `about:config` in Firefox and search for `webauthn`, you can see that the option is enabled, indicating that passwordless login is enabled.

If that isn't enough proof, go to https://webauthn.io/ and test it for yourself.

**Comment by ynojima** [comment]
It seems my understanding of "password-less" and your understanding of it is different.

In my understanding, "password-less" is "verifying user by biometrics or PIN code which is stored in authenticator locally". possession of authenticator is not enough authentication factor to verify who he is.

Your understanding seems possession of authenticator is enough to verify who he is.
Yes, in dictionary meaning, you are right. Firefox does support password-less login. They can verify possession of authenticator, which may be some sort of "password-less" authentication.
 but spring-security-webauthn is not designed to support to login with possession factor only since authenticator may be stolen.
spring-security-webauthn is designed to login with the combination of knowledge factor (password or PIN) and possession factor or the combination of biometrics factor (fingerprint) and possession factor.

**Comment by jajmo** [comment]
Ah, okay I see. In that case, you are indeed correct. Firefox and Chrome should have support for Windows Hello within the next few months.

**Comment by ynojima** [comment]
Yes, I am longing for it!

**Comment by ynojima** [comment]
Since it looks like we are on the same page now, I'll go ahead and close this. If you still have an issue, please reopen.
